File: src/models/account_tag.ts
export class AccountTag implements AccountTagInfoResponse {
    public id: string;
    public name: string;
    public displayOrder: number;
    public hidden: boolean;

    private constructor(id: string, name: string, displayOrder: number, hidden: boolean) {
        this.id = id;
        this.name = name;
        this.displayOrder = displayOrder;
        this.hidden = hidden;
    }

    public toCreateRequest(): AccountTagCreateRequest {
        return {
            name: this.name
        };
    }

    public toModifyRequest(): AccountTagModifyRequest {
        return {
            id: this.id,
            name: this.name
        };
    }

    public clone(): AccountTag {
        return new AccountTag(this.id, this.name, this.displayOrder, this.hidden);
    }

    public static of(tagResponse: AccountTagInfoResponse): AccountTag {
        return new AccountTag(tagResponse.id, tagResponse.name, tagResponse.displayOrder, tagResponse.hidden);
    }

    public static ofMulti(tagResponses: AccountTagInfoResponse[]): AccountTag[] {
        const tags: AccountTag[] = [];

        for (const tagResponse of tagResponses) {
            tags.push(AccountTag.of(tagResponse));
        }

        return tags;
    }

    public static createNewTag(name?: string): AccountTag {
        return new AccountTag('', name || '', 0, false);
    }
}

export interface AccountTagCreateRequest {
    readonly name: string;
}

export interface AccountTagModifyRequest {
    readonly id: string;
    readonly name: string;
}

export interface AccountTagHideRequest {
    readonly id: string;
    readonly hidden: boolean;
}

export interface AccountTagMoveRequest {
    readonly newDisplayOrders: AccountTagNewDisplayOrderRequest[];
}

export interface AccountTagNewDisplayOrderRequest {
    readonly id: string;
    readonly displayOrder: number;
}

export interface AccountTagDeleteRequest {
    readonly id: string;
}

export interface AccountTagInfoResponse {
    readonly id: string;
    readonly name: string;
    readonly displayOrder: number;
    readonly hidden: boolean;
}

File: pkg/models/account_tag.go
package models

// AccountTag represents account tag data stored in database
type AccountTag struct {
	TagId           int64  `xorm:"PK"`
	Uid             int64  `xorm:"INDEX(IDX_account_tag_uid_deleted_order) NOT NULL"`
	Deleted         bool   `xorm:"INDEX(IDX_account_tag_uid_deleted_order) NOT NULL"`
	Name            string `xorm:"VARCHAR(64) NOT NULL"`
	DisplayOrder    int32  `xorm:"INDEX(IDX_account_tag_uid_deleted_order) NOT NULL"`
	Hidden          bool   `xorm:"NOT NULL"`
	CreatedUnixTime int64
	UpdatedUnixTime int64
	DeletedUnixTime int64
}

// AccountTagGetRequest represents all parameters of account tag getting request
type AccountTagGetRequest struct {
	Id int64 `form:"id,string" binding:"required,min=1"`
}

// AccountTagCreateRequest represents all parameters of account tag creation request
type AccountTagCreateRequest struct {
	Name string `json:"name" binding:"required,notBlank,max=64"`
}

// AccountTagModifyRequest represents all parameters of account tag modification request
type AccountTagModifyRequest struct {
	Id   int64  `json:"id,string" binding:"required,min=1"`
	Name string `json:"name" binding:"required,notBlank,max=64"`
}

// AccountTagHideRequest represents all parameters of account tag hiding request
type AccountTagHideRequest struct {
	Id     int64 `json:"id,string" binding:"required,min=1"`
	Hidden bool  `json:"hidden"`
}

// AccountTagMoveRequest represents all parameters of account tag moving request
type AccountTagMoveRequest struct {
	NewDisplayOrders []*AccountTagNewDisplayOrderRequest `json:"newDisplayOrders" binding:"required,min=1"`
}

// AccountTagNewDisplayOrderRequest represents a data pair of id and display order
type AccountTagNewDisplayOrderRequest struct {
	Id           int64 `json:"id,string" binding:"required,min=1"`
	DisplayOrder int32 `json:"displayOrder"`
}

// AccountTagDeleteRequest represents all parameters of account tag deleting request
type AccountTagDeleteRequest struct {
	Id int64 `json:"id,string" binding:"required,min=1"`
}

// AccountTagInfoResponse represents a view-object of account tag
type AccountTagInfoResponse struct {
	Id           int64  `json:"id,string"`
	Name         string `json:"name"`
	DisplayOrder int32  `json:"displayOrder"`
	Hidden       bool   `json:"hidden"`
}

// FillFromOtherTag fills all the fields in this current tag from other account tag
func (t *AccountTag) FillFromOtherTag(tag *AccountTag) {
	t.TagId = tag.TagId
	t.Uid = tag.Uid
	t.Deleted = tag.Deleted
	t.Name = tag.Name
	t.DisplayOrder = tag.DisplayOrder
	t.Hidden = tag.Hidden
	t.CreatedUnixTime = tag.CreatedUnixTime
	t.UpdatedUnixTime = tag.UpdatedUnixTime
	t.DeletedUnixTime = tag.DeletedUnixTime
}

// ToAccountTagInfoResponse returns a view-object according to database model
func (t *AccountTag) ToAccountTagInfoResponse() *AccountTagInfoResponse {
	return &AccountTagInfoResponse{
		Id:           t.TagId,
		Name:         t.Name,
		DisplayOrder: t.DisplayOrder,
		Hidden:       t.Hidden,
	}
}

// AccountTagInfoResponseSlice represents the slice data structure of AccountTagInfoResponse
type AccountTagInfoResponseSlice []*AccountTagInfoResponse

// Len returns the count of items
func (s AccountTagInfoResponseSlice) Len() int {
	return len(s)
}

// Swap swaps two items
func (s AccountTagInfoResponseSlice) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

// Less reports whether the first item is less than the second one
func (s AccountTagInfoResponseSlice) Less(i, j int) bool {
	return s[i].DisplayOrder < s[j].DisplayOrder
}
